### PR TITLE
#22916 [CUBRID] Add Query Plan Feature for the CUBRID plugin

### LINKED ISSUE #22916: [CUBRID] Add Query Plan Feature for the CUBRID plugin
**Problem**
There is no Query Plan feature on the Cubrid plugin.

**Solution**
- Display query cost plan as table view
- Show raw query plan in property

### PR DESCRIPTION
**Pull request by rathana-pvs**
- Display query cost plan as table view
- Show raw query plan in property

### CODE REVIEW

**Comment by LonwoLonwo** [comment]
Hello @rathana-pvs 

Thanks for your contribution.
We will look at it.

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                return "Index Join - Inner Join";
+            case "m-join (inner join)":
+                return "Merged - Inner Join";
+            case "temp":
+                return "Temp";
+            case "follow":
+                return "Follow";
+            default:
+                return method;
+        }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java]: Switch expression can be used instead:

```suggestion
        return switch (method) {
            case "iscan" -> "Index Scan";
            case "sscan" -> "Full Scan";
            case "temp(group by)" -> "Group by Temp";
            case "temp(order by)" -> "Order by Temp";
            case "nl-join (inner join)" -> "Nested Loop - Inner Join";
            case "nl-join (cross join)" -> "Nested Loop - Cross Join";
            case "idx-join (inner join)" -> "Index Join - Inner Join";
            case "m-join (inner join)" -> "Merged - Inner Join";
            case "temp" -> "Temp";
            case "follow" -> "Follow";
            default -> method;
        };
```

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    private Map<String, String> nodeProps = new HashMap<>();
+
+    private CubridPlanNode parent;
+    private List<CubridPlanNode> nested;
+
+    public CubridPlanNode(String queryPlan) 
+    {
+        this.fullText = queryPlan;
+        parseObject(this.getSegments());
+        parseNode();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java]: Can delegate to the overloaded constructor

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        return parent;
+    }
+
+    @Override
+    public Collection<CubridPlanNode> getNested() 
+    {
+        return nested;
+    }
+
+    public String getMethodTitle(String method)
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java]: Can be `private`

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    private CubridPlanNode(CubridPlanNode parent, String name, List<String> segments, String fullText) 
+    {
+        this.parent = parent;
+        this.name = name;
+        this.fullText = fullText;
+        parseObject(segments);
+        parseNode();
+    }
+
+    private void parseObject(List<String> segments)
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java]: Please place all private methods at the bottom of the class

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridStatementProxy.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            if (e.getTargetException() instanceof SQLException) {
+                throw new SQLException(e.getMessage(), e.getTargetException());
+            } else {
+                throw new SQLException(
+                        e.getMessage() + "\r\n" + e.getTargetException().getMessage(),
+                        null,
+                        -90002);
+            }
+        }
+    }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridStatementProxy.java]: Please use `org.jkiss.utils.BeanUtils#invokeObjectMethod(java.lang.Object, java.lang.String)` instead

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridStatementProxy.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ * limitations under the License.
+ */ 
+package org.jkiss.dbeaver.ext.cubrid.model.plan;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+public class CubridStatementProxy
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridStatementProxy.java]: There is no need to create a wrapper to call a function on the underlying object. Make a static method instead.

**Review by ShadelessFox — COMMENTED** [review]
Please format your code according to our code style rules:
- Place `{` at the same line where the method is declared
- Use `@NotNull`/`@Nullable` annotations for method arguments and return type

File: plugins/org.jkiss.dbeaver.ext.cubrid/OSGI-INF/l10n/bundle.properties [hunk 7]
```diff
@@ -44,3 +44,7 @@ meta.org.jkiss.dbeaver.ext.cubrid.model.CubridUser.name.name=Name
 meta.org.jkiss.dbeaver.ext.cubrid.model.CubridUser.comment.name=Comment
 meta.org.jkiss.dbeaver.ext.cubrid.model.CubridTable.schema.name=Owner
 meta.org.jkiss.dbeaver.ext.cubrid.model.CubridView.schema.name=Owner
+meta.org.jkiss.dbeaver.ext.cubrid.model.plan.CubridPlanNode.cardinality.name=Cardinality
+meta.org.jkiss.dbeaver.ext.cubrid.model.plan.CubridPlanNode.fullText.name=FullText
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.cubrid/OSGI-INF/l10n/bundle.properties]:
```suggestion
meta.org.jkiss.dbeaver.ext.cubrid.model.plan.CubridPlanNode.fullText.name=Full Text

```

**Comment by rathana-pvs** [comment]
> Please format your code according to our code style rules:
> 
> * Place `{` at the same line where the method is declared
> * Use `@NotNull`/`@Nullable` annotations for method arguments and return type

Will it apply to the entire Cubrid plugin?

**Comment by LonwoLonwo** [comment]
> Will it apply to the entire Cubrid plugin?

Hello @rathana-pvs 

This advice is only for your code.

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                String[] values = nodeProps.get(key).split(" card ");
+                this.cost = Integer.parseInt(values[0]);
+                this.row = Integer.parseInt(values[1]);
+            }
+        }
+    }
+
+    private void parseObject(@NotNull List<String> segments) {
+        if (!segments.isEmpty()) {
+            String[] removes = segments.remove(0).split(":");
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java]: Could you please move all hardcoded strings with duplicates into constants?
Like the:
`private static final String OPTIONS_SEPARATOR = ":";`

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+                parseObject(segments);
+            }
+        }
+    }
+    
+    @NotNull
+    private List<String> getSegments() {
+        Pattern pattern =
+                Pattern.compile(
+                        "[\\n\\r]node\\[....\\s*([^\\n\\r]*)|[\\n\\r].*Query plan:\\s*([^\\n\\r]*)|[\\n\\r].*subplan:\\s*([^\\n\\r]*)|[\\n\\r].*cost:\\s*([^\\n\\r]*)|inner:\\s*([^\\n\\r]*)|outer:\\s*([^\\n\\r]*)|class:\\s*([^\\n\\r]*)");
```
Comment (Destrolaric) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/plan/CubridPlanNode.java]: I am not sure about this regex, I tested it, and it is recognized pretty badly. Can it be separated into simpler regexes?

**Review by Destrolaric — CHANGES_REQUESTED** [review]

**Review by ShadelessFox — APPROVED** [review]

**Review by Destrolaric — APPROVED** [review]

**Comment by rathana-pvs** [comment]
What else do I need to do to get this PR merged into devel?

**Comment by ShadelessFox** [comment]
Hello @rathana-pvs,

It awaits our QA team. Once it passes the QA, it will be merged.

**Comment by uslss** [comment]
verified

**Pull request merged**